src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+class GlobalFieldCacher {
+public:
+  GlobalFieldCacher(MultiStage* msptr, StencilInstantiation* si)
+      : multiStagePrt_(msptr), instantiation_(si) {}
+
+  /// @brief entry method for the pass: processes a given multistage and applies all changes
+  /// required
+  void process() {
+    computeOptimalFields();
+    addFillerStages();
+    addFlushStages();
+  }
+
+  std::vector<nameToImprovementMetric> getoriginalNameToCache() { return originalNameToCache_; }
+
+private:
+  /// @brief use the data locality metric to rank the fields in a stencil based on how much they
+  /// would benefit from caching
+  void computeOptimalFields() {
+    auto dataLocality =
+        computeReadWriteAccessesMetricIndividually(instantiation_, *multiStagePrt_, config_);
+
+    for(const auto& stagePtr : multiStagePrt_->getStages()) {
+      for(const Field& field : stagePtr->getFields()) {
+        auto findField = accessIDtoDataLocality_.find(field.AccessID);
+
+        // We already checked this field
+        if(findField != accessIDtoDataLocality_.end())
+          continue;
+
+        // Only ij-cache fields that are not horizontally pointwise and are vertically pointwise
+        if(!(field.Extent.isVerticalPointwise()) || (field.Extent.isHorizontalPointwise()))
+          continue;
+
+        // This is caching non-temporary fields
+        if(instantiation_->isTemporaryField(field.AccessID))
+          continue;
+
+        int cachedreadwrites = dataLocality.find(field.AccessID)->second.total();
+
+        // We check how much the cache-filling costs: 1 Read and if we fill from Memory
+        if(checkReadBeforeWrite(field.AccessID))
+          cachedreadwrites--;
+        // We need one more write to flush the cache back to the variable
+        if(!checkReadOnlyAccess(field.AccessID))
+          cachedreadwrites--;
+
+        if(cachedreadwrites > 0)
+          accessIDtoDataLocality_.emplace(field.AccessID, cachedreadwrites);
+      }
+    }
+
+    for(auto& performanceMesure : accessIDtoDataLocality_) {
+      sortedAccesses_.emplace_back(performanceMesure.first, performanceMesure.second);
+      std::sort(sortedAccesses_.begin(), sortedAccesses_.end());
+    }
+  }
+
+  /// @brief for each chached variable, check if we need a filler-stage (if we find a
+  /// read-before-write) and add it if necessary, renames all occurances and sets the cache for the
+  /// temporary variable newly created
+  ///
+  /// We always fill to the extent of the given multistage and we only add one stage to fill all the
+  /// variables to reduce synchronisation overhead
+  void addFillerStages() {
+    for(int i = 0; i < std::min((int)sortedAccesses_.size(), config_.SMemMaxFields); ++i) {
+      int oldID = sortedAccesses_[i].accessID;
+
+      // Create new temporary field and register in the instantiation
+      int newID = instantiation_->nextUID();
+
+      instantiation_->setAccessIDNamePairOfField(newID, dawn::format("__tmp_cache_%i", i), true);
```

Comment:
`instantiation_->setAccessIDNamePairOfField(newID, "__tmp_cache_" + std::to_string(i), true);` [this an order of magnitude faster ;)]